File: static/js/explanations/builders/osno.js
(function (global) {
    const format = global.ExplanationsFormat;
    if (!format) {
        return;
    }
    const { fmtMoney, fmtPercent, renderMarkdown } = format;

    function buildOsnoMarkdown(params) {
        const revenue = params.revenue || 0;
        const tax = params.tax || 0;
        const vatPayment = params.vat || 0;
        const insurance = params.insurance || 0;
        const totalBurden = params.totalBurden || 0;
        const burdenPct = params.burdenPct || 0;
        const baseCost = params.baseCost || 0;
        const baseRent = params.baseRent || 0;
        const baseOther = params.baseOther || 0;
        const baseFot = params.baseFot || 0;
        const baseInsurStd = params.baseInsurStd || 0;
        const incomeWithoutVat = params.incomeWithoutVat || 0;
        const expensesWithoutVatParam = params.expensesWithoutVat || 0;
        const vatPurchasesPercent = params.vatPurchasesPercent || 0;
        const rentSharePercent = (params.vatShareRent || 0) * 100;
        const otherSharePercent = (params.vatShareOther || 0) * 100;
        const stockExtra = params.stockExtra || 0;
        const accumulatedVatCredit = params.accumulatedVatCredit || 0;
        const transitionMode = params.transitionMode || 'none';
        const vatChargedParam = params.vatCharged;
        const vatDeductibleParam = params.vatDeductible;
        const vatPayableBalance = params.vatPayable !== undefined
            ? params.vatPayable
            : (params.netProfitAccounting || 0) - (params.netProfitCash || params.netProfit || 0);
        const vatRefund = params.vatRefund || 0;
        const cogsVat = params.vatDeductibleCogs || 0;
        const rentVat = params.vatDeductibleRent || 0;
        const otherVat = params.vatDeductibleOther || 0;
        const cogsNoVat = params.cogsNoVat || Math.max(baseCost - cogsVat, 0);
        const rentNoVat = params.rentNoVat || Math.max(baseRent - rentVat, 0);
        const otherNoVat = params.otherNoVat || Math.max(baseOther - otherVat, 0);
        const netProfitAccounting = params.netProfitAccounting !== undefined
            ? params.netProfitAccounting
            : ((incomeWithoutVat || 0) - expensesWithoutVatParam - tax);
        const netProfitCash = params.netProfitCash !== undefined
            ? params.netProfitCash
            : (netProfitAccounting - vatPayableBalance);
        const expensesWithoutVat = expensesWithoutVatParam
            || (cogsNoVat + rentNoVat + otherNoVat + baseFot + baseInsurStd + (stockExtra > 0 ? stockExtra : 0));
        const vatCharged = vatChargedParam || (revenue * 22 / 122);
        const vatDeductible = vatDeductibleParam || (cogsVat + rentVat + otherVat);
        const creditPart = transitionMode === 'vat' && accumulatedVatCredit > 0
            ? ` − ${fmtMoney(accumulatedVatCredit)} ₽ (накопленный НДС)`
            : '';
        const revenueNet = incomeWithoutVat > 0 ? incomeWithoutVat : Math.max(revenue - vatCharged, 0);
        const profitBase = revenueNet - expensesWithoutVat;
        const positiveProfitBase = Math.max(profitBase, 0);
        const burdenPercentText = fmtPercent(burdenPct);
        const stockPart = stockExtra > 0 ? stockExtra : 0;

        const lines = [
            '# Пояснение к расчёту налоговой нагрузки (ОСНО + НДС 22% (ООО))',
            '',
            '## 1. Вводные данные',
            'Режим: ОСНО + НДС 22% (ООО).',
            '',
            'Учитывается:',
            '- НДС 22%: начисленный с выручки минус входящий по закупкам (себестоимость, аренда, прочие).',
            '- Налог на прибыль 25% с базы после расходов без НДС (ФОТ и взносы входят целиком).',
            '- Страховые взносы: только 30% от ФОТ (для ООО нет взносов ИП).',
            '',
            '## 2. Доходы',
            `Выручка с НДС: ${fmtMoney(revenue)} ₽.`,
            'НДС, включённый в цену (метод «включённый НДС»):',
            `${fmtMoney(revenue)} × 22 / 122 = ${fmtMoney(vatCharged)} ₽.`,
            `Доходы без НДС (база для налога на прибыль): ${fmtMoney(revenueNet)} ₽.`,
            '',
            '## 3. Расходы для налога на прибыль (без НДС)',
            'Себестоимость, аренда и прочие расходы вводятся с НДС. Для базы по прибыли берём суммы без НДС.',
            `Итого расходы без НДС: ${fmtMoney(expensesWithoutVat)} ₽.`,
            '',
            'Структура расходов:',
            `- Себестоимость: ${fmtMoney(baseCost)} ₽ с НДС → ${fmtMoney(cogsNoVat)} ₽ без НДС (входящий НДС ${fmtMoney(cogsVat)} ₽, доля облагаемых закупок ${fmtPercent(vatPurchasesPercent)}%).`,
            `- Аренда: ${fmtMoney(baseRent)} ₽ с НДС → ${fmtMoney(rentNoVat)} ₽ без НДС (входящий НДС ${fmtMoney(rentVat)} ₽, доля с НДС ${fmtPercent(rentSharePercent)}%).`,
            `- Прочие: ${fmtMoney(baseOther)} ₽ с НДС → ${fmtMoney(otherNoVat)} ₽ без НДС (входящий НДС ${fmtMoney(otherVat)} ₽, доля с НДС ${fmtPercent(otherSharePercent)}%).`,
            `- ФОТ (без НДС): ${fmtMoney(baseFot)} ₽.`,
            `- Страховые взносы 30% от ФОТ: ${fmtMoney(baseInsurStd)} ₽.`,
            ...(stockPart > 0 ? [`- Списание остатков при переходе: ${fmtMoney(stockPart)} ₽.`] : []),
            '',
            '## 4. Расчёт налога на прибыль 25%',
            'Налог на прибыль = (доходы без НДС − расходы без НДС) × 25%.',
            'Налоговая база:',
            `${fmtMoney(revenueNet)} − ${fmtMoney(expensesWithoutVat)} = ${fmtMoney(profitBase)} ₽.`,
            '',
            'Налог на прибыль:',
            `${fmtMoney(positiveProfitBase)} × 25% = ${fmtMoney(tax)} ₽.`,
            '',
            '## 5. Расчёт НДС 22%',
            '',
            '### 5.1. НДС к начислению',
            'Формула та же: выручка × 22 / 122.',
            `${fmtMoney(revenue)} × 22 / 122 = ${fmtMoney(vatCharged)} ₽.`,
            '',
            '### 5.2. Входящий НДС к вычету',
            `- Себестоимость: ${fmtMoney(baseCost)} ₽ × ${fmtPercent(vatPurchasesPercent)}% × 22 / 122 = ${fmtMoney(cogsVat)} ₽.`,
            `- Аренда: ${fmtMoney(baseRent)} ₽ × ${fmtPercent(rentSharePercent)}% × 22 / 122 = ${fmtMoney(rentVat)} ₽.`,
            `- Прочие расходы: ${fmtMoney(baseOther)} ₽ × ${fmtPercent(otherSharePercent)}% × 22 / 122 = ${fmtMoney(otherVat)} ₽.`,
            `Итого входящий НДС: ${fmtMoney(vatDeductible)} ₽.`,
        ];

        if (transitionMode === 'vat' && accumulatedVatCredit > 0) {
            lines.push(
                '',
                `Дополнительный вычет при переходе: накопленный входящий НДС = ${fmtMoney(accumulatedVatCredit)} ₽.`,
            );
        }

        lines.push(
            '',
            '### 5.3. НДС к уплате',
            'Расчёт НДС к уплате (баланс):',
            `${fmtMoney(vatCharged)} − ${fmtMoney(vatDeductible)}${creditPart} = ${fmtMoney(vatPayableBalance)} ₽.`,
        );

        if (vatPayableBalance < 0 && vatRefund > 0) {
            lines.push(`Получилась сумма к возмещению: ${fmtMoney(vatRefund)} ₽.`);
        } else {
            lines.push(`Фактический платёж в бюджет: ${fmtMoney(vatPayment)} ₽.`);
        }

        lines.push(
            '',
            '## 6. Страховые взносы',
            'ООО платит только взносы 30% с ФОТ:',
            `${fmtMoney(baseFot)} × 30% = ${fmtMoney(baseInsurStd)} ₽.`,
            '',
            'Итого страховых взносов:',
            `${fmtMoney(baseInsurStd)} ₽ = ${fmtMoney(insurance)} ₽.`,
            '',
            '## 7. Совокупные платежи и налоговая нагрузка',
            'Платежи = налог на прибыль + НДС к перечислению + страховые взносы.',
            `${fmtMoney(tax)} + ${fmtMoney(vatPayment)} + ${fmtMoney(insurance)} = ${fmtMoney(totalBurden)} ₽.`,
            '',
            'Доля от выручки:',
            `${fmtMoney(totalBurden)} / ${fmtMoney(revenue)} × 100% = ${burdenPercentText}%.`,
            '',
            '## 8. Чистая прибыль',
            '- Бухгалтерская (P&L, без учёта НДС к уплате):',
            `${fmtMoney(revenueNet)} − ${fmtMoney(expensesWithoutVat)} − ${fmtMoney(tax)} = ${fmtMoney(netProfitAccounting)} ₽.`,
            '',
            '- Денежная (cash, с учётом НДС как денежного потока):',
            `${fmtMoney(netProfitAccounting)} − ${fmtMoney(vatPayableBalance)} = ${fmtMoney(netProfitCash)} ₽.`,
        );

        return renderMarkdown(lines);
    }

    function buildOsnoIpMarkdown(params) {
        const revenue = params.revenue || 0;
        const vat = params.vat || 0;
        const insurance = params.insurance || 0;
        const totalBurden = params.totalBurden || 0;
        const burdenPct = params.burdenPct || 0;
        const baseCost = params.baseCost || 0;
        const baseRent = params.baseRent || 0;
        const baseOther = params.baseOther || 0;
        const baseFot = params.baseFot || 0;
        const baseInsurStd = params.baseInsurStd || 0;
        const fixedContrib = params.fixedContrib || 0;
        const ownerExtra = params.ownerExtra || 0;
        const ownerExtraBase = params.ownerExtraBase || 0;
        const vatPurchasesPercent = params.vatPurchasesPercent || 0;
        const stockExtra = params.stockExtra || 0;
        const accumulatedVatCredit = params.accumulatedVatCredit || 0;
        const transitionMode = params.transitionMode || 'none';
        const incomeWithoutVat = params.incomeWithoutVat || 0;
        const expensesWithoutVat = params.expensesWithoutVat || 0;
        const ndflTax = params.ndflTax || params.tax || 0;

        const stockPart = stockExtra > 0 ? stockExtra : 0;
        const vatRate = 22;
        const vatCharged = params.vatCharged || (revenue * vatRate / (100 + vatRate));
        const cogsVat = params.vatDeductibleCogs || (baseCost * (vatPurchasesPercent / 100) * vatRate / (100 + vatRate));
        const rentVat = params.vatDeductibleRent || 0;
        const otherVat = params.vatDeductibleOther || 0;
        const cogsNoVat = params.cogsNoVat || Math.max(baseCost - cogsVat, 0);
        const rentNoVat = params.rentNoVat || Math.max(baseRent - rentVat, 0);
        const otherNoVat = params.otherNoVat || Math.max(baseOther - otherVat, 0);
        const vatDeductible = params.vatDeductible || (cogsVat + rentVat + otherVat);
        const vatPayableBalance = params.vatPayable !== undefined
            ? params.vatPayable
            : (params.netProfitAccounting || (incomeWithoutVat - expensesWithoutVat - ndflTax)) - (params.netProfitCash || params.netProfit || 0);
        const vatRefund = params.vatRefund || 0;
        const rentSharePercent = (params.vatShareRent || 0) * 100;
        const otherSharePercent = (params.vatShareOther || 0) * 100;
        const netProfitAccounting = params.netProfitAccounting !== undefined
            ? params.netProfitAccounting
            : (incomeWithoutVat - expensesWithoutVat - ndflTax);
        const netProfitCash = params.netProfitCash !== undefined
            ? params.netProfitCash
            : (netProfitAccounting - vatPayableBalance);
        const creditPart = transitionMode === 'vat' && accumulatedVatCredit > 0
            ? ` − ${fmtMoney(accumulatedVatCredit)} ₽ (накопленный НДС)`
            : '';
        const baseWithout1pct = Math.max(incomeWithoutVat - expensesWithoutVat - fixedContrib, 0);
        const excessOverThreshold = Math.max(baseWithout1pct - 300000, 0);
        const extra1pctCalc = ownerExtra || excessOverThreshold * 0.01;
        const ndflBaseCalc = Math.max(baseWithout1pct - extra1pctCalc, 0);
        const insuranceBreakdown = `${fmtMoney(baseInsurStd)} ₽ + ${fmtMoney(extra1pctCalc)} ₽ + ${fmtMoney(fixedContrib)} ₽`;
        const burdenPercentText = fmtPercent(burdenPct);
        const ndflEffective = ndflBaseCalc > 0 ? (ndflTax / ndflBaseCalc * 100) : 0;
        const ndflScale = [
            'до 2 400 000 ₽ — 13%',
            '2 400 001–5 000 000 ₽ — 15%',
            '5 000 001–20 000 000 ₽ — 18%',
            '20 000 001–50 000 000 ₽ — 20%',
            'свыше 50 000 000 ₽ — 22%',
        ];

        const lines = [
            '# Пояснение к расчёту налоговой нагрузки (ОСНО + НДС 22% (ИП))',
            '',
            '## 1. Вводные данные',
            'Режим: ОСНО + НДС 22% (ИП).',
            '',
            'Учитывается:',
            '- НДС рассчитывается так же, как у юридических лиц: начисление с выручки и вычеты по закупкам.',
            '- Вместо налога на прибыль рассчитывается НДФЛ ИП по прогрессивной шкале.',
            '- В налоговую базу по НДФЛ включены расходы без НДС и фиксированный взнос ИП; дополнительный 1% рассчитывается отдельно.',
            '',
            '## 2. Доходы',
            `Выручка (с НДС): ${fmtMoney(revenue)} ₽.`,
            `Доходы без НДС для расчёта НДФЛ: ${fmtMoney(incomeWithoutVat)} ₽.`,
            '',
            '## 3. Расходы (без НДС)',
            `Расходы, уменьшающие базу НДФЛ: ${fmtMoney(expensesWithoutVat)} ₽.`,
            '',
            'Структура расходов:',
            `- Себестоимость (без НДС по облагаемым закупкам): ${fmtMoney(cogsNoVat)} ₽.`,
            `- Аренда (без НДС): ${fmtMoney(rentNoVat)} ₽.`,
            `- Прочие расходы (без НДС): ${fmtMoney(otherNoVat)} ₽.`,
            `- ФОТ: ${fmtMoney(baseFot)} ₽.`,
            `- Страховые взносы 30% от ФОТ: ${fmtMoney(baseInsurStd)} ₽.`,
            ...(stockPart > 0 ? [`- Списание остатков товара: ${fmtMoney(stockPart)} ₽.`] : []),
            '',
            '## 4. База ИП и взнос 1% свыше 300 000 ₽',
            '1. База ИП без учёта 1%:',
            `   ${fmtMoney(incomeWithoutVat)} − ${fmtMoney(expensesWithoutVat)} − ${fmtMoney(fixedContrib)} = ${fmtMoney(baseWithout1pct)} ₽.`,
            ...(excessOverThreshold > 0
                ? [
                    '',
                    '2. Превышение над 300 000 ₽:',
                    `   ${fmtMoney(baseWithout1pct)} − 300 000 ₽ = ${fmtMoney(excessOverThreshold)} ₽.`,
                    '',
                    '3. Дополнительный взнос 1%:',
                    `   ${fmtMoney(excessOverThreshold)} × 1% = ${fmtMoney(extra1pctCalc)} ₽.`,
                    '',
                    '4. База для НДФЛ:',
                    `   ${fmtMoney(baseWithout1pct)} − ${fmtMoney(extra1pctCalc)} = ${fmtMoney(ndflBaseCalc)} ₽.`,
                    '',
                ]
                : [
                    '',
                    '2. База не превышает 300 000 ₽, взнос 1% = 0 ₽.',
                    '',
                    '3. База для НДФЛ:',
                    `   ${fmtMoney(baseWithout1pct)} = ${fmtMoney(ndflBaseCalc)} ₽.`,
                    '',
                ]),
            '## 5. Расчёт НДФЛ по прогрессивной шкале',
            `Ступени (2026): ${ndflScale.join(', ')}.`,
            '',
            `Итоговый НДФЛ: ${fmtMoney(ndflTax)} ₽ (эффективная ставка ${fmtPercent(ndflEffective)}%).`,
            '',
            '## 6. Расчёт НДС 22%',
            '',
            '### 6.1. НДС к начислению',
            `${fmtMoney(revenue)} × 22 / 122 = ${fmtMoney(vatCharged)} ₽.`,
            '',
            '### 6.2. НДС к вычету по закупкам',
            `- Себестоимость: ${fmtMoney(baseCost)} ₽ × ${fmtPercent(vatPurchasesPercent)}% × 22 / 122 = ${fmtMoney(cogsVat)} ₽.`,
            `- Аренда: ${fmtMoney(baseRent)} ₽ × ${fmtPercent(rentSharePercent)}% × 22 / 122 = ${fmtMoney(rentVat)} ₽.`,
            `- Прочие расходы: ${fmtMoney(baseOther)} ₽ × ${fmtPercent(otherSharePercent)}% × 22 / 122 = ${fmtMoney(otherVat)} ₽.`,
            `Итого входящий НДС: ${fmtMoney(vatDeductible)} ₽.`,
        ];

        if (transitionMode === 'vat' && accumulatedVatCredit > 0) {
            lines.push(
                '',
                `Дополнительный вычет при переходе: накопленный входящий НДС = ${fmtMoney(accumulatedVatCredit)} ₽.`,
            );
        }

        lines.push(
            '',
            '### 6.3. НДС к уплате',
            `${fmtMoney(vatCharged)} − ${fmtMoney(vatDeductible)}${creditPart} = ${fmtMoney(vatPayableBalance)} ₽.`,
            vatPayableBalance < 0 && vatRefund > 0
                ? `Сумма к возмещению: ${fmtMoney(vatRefund)} ₽.`
                : `Платёж в бюджет: ${fmtMoney(vat)} ₽.`,
            '',
            '## 7. Страховые взносы',
            '30% от ФОТ:',
            `${fmtMoney(baseFot)} × 30% = ${fmtMoney(baseInsurStd)} ₽.`,
            '',
            '1% с доходов свыше 300 000 ₽:',
            ...(excessOverThreshold > 0
                ? [`   ${fmtMoney(excessOverThreshold)} × 1% = ${fmtMoney(extra1pctCalc)} ₽.`]
                : ['   превышения нет, взнос 1% = 0 ₽.']),
            '',
            'Фиксированный взнос ИП:',
            `${fmtMoney(fixedContrib)} ₽.`,
            '',
            'Итого страховых взносов:',
            `${insuranceBreakdown} = ${fmtMoney(insurance)} ₽.`,
            '',
            '## 8. Совокупная налоговая нагрузка',
            'Сумма налогов и взносов:',
            `${fmtMoney(ndflTax)} + ${fmtMoney(vat)} + ${fmtMoney(insurance)} = ${fmtMoney(totalBurden)} ₽.`,
            '',
            'Доля от выручки:',
            `${fmtMoney(totalBurden)} / ${fmtMoney(revenue)} × 100% = ${burdenPercentText}%.`,
            '',
            '## 9. Чистая прибыль',
            '- Бухгалтерская (до НДС к уплате):',
            `${fmtMoney(incomeWithoutVat)} − ${fmtMoney(expensesWithoutVat)} − ${fmtMoney(ndflTax)} = ${fmtMoney(netProfitAccounting)} ₽.`,
            '',
            '- Денежная (cash) с учётом движения по НДС:',
            `${fmtMoney(netProfitAccounting)} − ${fmtMoney(vatPayableBalance)} = ${fmtMoney(netProfitCash)} ₽.`,
            );

        return renderMarkdown(lines);
    }

    global.ExplanationBuilders = global.ExplanationBuilders || {};
    global.ExplanationBuilders.buildOsnoMarkdown = buildOsnoMarkdown;
    global.ExplanationBuilders.buildOsnoIpMarkdown = buildOsnoIpMarkdown;
})(window);

File: calculator/regimes/osno.py
from __future__ import annotations

from typing import Optional, Tuple

from ..constants import PROFIT_TAX_RATE, VAT_RATE_STANDARD, THRESHOLD_1_PERCENT
from ..insurance import calculate_progressive_ndfl
from ..models import CalcInput, CalcResult, CalculationContext
from ..vat import calc_vat_charged, calc_vat_to_pay


def _normalize_share(value: Optional[float], default: float) -> float:
    """Ensure VAT share stays within 0..1 bounds."""
    if value is None:
        share = default
    else:
        share = value
    if share > 1.0:
        share = share / 100.0
    if share < 0.0:
        return 0.0
    return min(share, 1.0)


def _split_amount_with_vat(amount: float, share_with_vat: float, vat_rate: float) -> Tuple[float, float]:
    """Return VAT portion and net (без НДС) amount for an expense."""
    if amount <= 0.0 or vat_rate <= 0.0 or share_with_vat <= 0.0:
        return 0.0, max(amount, 0.0)
    taxable_amount = amount * share_with_vat
    vat_amount = calc_vat_charged(taxable_amount, vat_rate)
    net_amount = amount - vat_amount
    return vat_amount, max(net_amount, 0.0)


def _build_expense_breakdown(data: CalcInput, ctx: CalculationContext, vat_rate: float):
    stock_part = ctx.stock_extra if ctx.stock_extra > 0 else 0.0
    payroll = ctx.annual_fot
    insurance = ctx.insurance_standard

    cogs_share = _normalize_share(getattr(data, "vat_share_cogs", None), data.vat_purchases_percent)
    rent_share = _normalize_share(getattr(data, "vat_share_rent", None), 1.0)
    other_share = _normalize_share(getattr(data, "vat_share_other", None), 1.0)

    cogs_vat, cogs_net = _split_amount_with_vat(ctx.cost_of_goods, cogs_share, vat_rate)
    rent_vat, rent_net = _split_amount_with_vat(data.rent, rent_share, vat_rate)
    other_vat, other_net = _split_amount_with_vat(ctx.other_expenses, other_share, vat_rate)

    expenses_without_vat = cogs_net + rent_net + other_net + payroll + insurance + stock_part
    vat_deductible_total = cogs_vat + rent_vat + other_vat

    return {
        "stock_part": stock_part,
        "payroll": payroll,
        "insurance": insurance,
        "cogs_net": cogs_net,
        "rent_net": rent_net,
        "other_net": other_net,
        "cogs_vat": cogs_vat,
        "rent_vat": rent_vat,
        "other_vat": other_vat,
        "expenses_without_vat": expenses_without_vat,
        "vat_deductible_total": vat_deductible_total,
    }


def calculate_osno_ooo(data: CalcInput, ctx: CalculationContext) -> CalcResult:
    vat_rate = VAT_RATE_STANDARD
    vat_charged = calc_vat_charged(data.revenue, vat_rate)
    expenses_info = _build_expense_breakdown(data, ctx, vat_rate)
    vat_deductible = expenses_info["vat_deductible_total"]

    vat_payable_balance = vat_charged - vat_deductible - ctx.vat_credit_to_apply
    vat_to_pay = calc_vat_to_pay(vat_charged, vat_deductible, ctx.vat_credit_to_apply)

    revenue_without_vat = data.revenue - vat_charged
    expenses_without_vat = expenses_info["expenses_without_vat"]

    profit_tax_base = revenue_without_vat - expenses_without_vat
    profit_tax = max(profit_tax_base, 0.0) * PROFIT_TAX_RATE
    net_profit_accounting = profit_tax_base - profit_tax
    net_profit_cash = net_profit_accounting - vat_payable_balance

    insurance = expenses_info["insurance"]
    total_payments = profit_tax + vat_to_pay + insurance

    extra = {
        "income_without_vat": revenue_without_vat,
        "expenses_without_vat": expenses_without_vat,
        "profit_tax_base": profit_tax_base,
        "vat_charged": vat_charged,
        "vat_deductible": vat_deductible,
        "vat_extra_credit": ctx.vat_credit_to_apply,
        "vat_payable": vat_payable_balance,
        "vat_refund": max(-vat_payable_balance, 0.0),
        "cogs_no_vat": expenses_info["cogs_net"],
        "rent_no_vat": expenses_info["rent_net"],
        "other_no_vat": expenses_info["other_net"],
        "vat_deductible_cogs": expenses_info["cogs_vat"],
        "vat_deductible_rent": expenses_info["rent_vat"],
        "vat_deductible_other": expenses_info["other_vat"],
        "net_profit_accounting": net_profit_accounting,
        "net_profit_cash": net_profit_cash,
        "total_payments": total_payments,
    }

    return CalcResult(
        regime="osno_ooo",
        title="ОСНО + НДС 22% (ООО)",
        revenue=data.revenue,
        expenses=expenses_without_vat,
        tax=profit_tax,
        vat=vat_to_pay,
        insurance=insurance,
        total_burden=total_payments,
        burden_percent=(total_payments / data.revenue * 100) if data.revenue > 0 else 0.0,
        net_profit=net_profit_cash,
        extra=extra,
    )


def calculate_osno_ip(data: CalcInput, ctx: CalculationContext) -> CalcResult:
    vat_rate = VAT_RATE_STANDARD
    vat_charged = calc_vat_charged(data.revenue, vat_rate)
    expenses_info = _build_expense_breakdown(data, ctx, vat_rate)
    vat_deductible = expenses_info["vat_deductible_total"]
    vat_payable_balance = vat_charged - vat_deductible - ctx.vat_credit_to_apply
    vat_to_pay = calc_vat_to_pay(vat_charged, vat_deductible, ctx.vat_credit_to_apply)

    income_without_vat = data.revenue - vat_charged
    business_expenses_without_vat = expenses_info["expenses_without_vat"]

    base_for_one_percent = income_without_vat - business_expenses_without_vat - data.fixed_contrib
    base_for_one_percent = max(base_for_one_percent, 0.0)
    extra_one_percent = max(base_for_one_percent - THRESHOLD_1_PERCENT, 0.0) * 0.01

    ndfl_base = base_for_one_percent - extra_one_percent
    ndfl_base = max(ndfl_base, 0.0)
    ndfl_tax = calculate_progressive_ndfl(ndfl_base)

    insurance_total = expenses_info["insurance"] + data.fixed_contrib + extra_one_percent
    total_tax_burden = ndfl_tax + vat_to_pay + insurance_total

    net_profit_accounting = income_without_vat - business_expenses_without_vat - ndfl_tax
    net_profit_cash = net_profit_accounting - vat_payable_balance

    extra = {
        "ndfl_base": ndfl_base,
        "income_without_vat": income_without_vat,
        "expenses_without_vat": business_expenses_without_vat,
        "fixed_contrib": data.fixed_contrib,
        "owner_extra": extra_one_percent,
        "owner_extra_base": base_for_one_percent,
        "vat_charged": vat_charged,
        "vat_deductible": vat_deductible,
        "vat_extra_credit": ctx.vat_credit_to_apply,
        "ndfl_tax": ndfl_tax,
        "vat_payable": vat_payable_balance,
        "vat_refund": max(-vat_payable_balance, 0.0),
        "cogs_no_vat": expenses_info["cogs_net"],
        "rent_no_vat": expenses_info["rent_net"],
        "other_no_vat": expenses_info["other_net"],
        "vat_deductible_cogs": expenses_info["cogs_vat"],
        "vat_deductible_rent": expenses_info["rent_vat"],
        "vat_deductible_other": expenses_info["other_vat"],
        "net_profit_accounting": net_profit_accounting,
        "net_profit_cash": net_profit_cash,
        "total_payments": total_tax_burden,
    }

    return CalcResult(
        regime="osno_ip",
        title="ОСНО + НДС 22% (ИП)",
        revenue=data.revenue,
        expenses=business_expenses_without_vat,
        tax=ndfl_tax,
        vat=vat_to_pay,
        insurance=insurance_total,
        total_burden=total_tax_burden,
        burden_percent=(total_tax_burden / data.revenue * 100) if data.revenue > 0 else 0.0,
        net_profit=net_profit_cash,
        extra=extra,
    )
